File: Week_06/java/src/longestpalindromicsubstring/_647_PalindromicSubstrings.java
package longestpalindromicsubstring;

/**
 * 第一遍：2020/08/19周三 ✅
 * 第二遍：2020/08/20周四 ✅
 * 第三遍：2020/07/08周四
 * 第四遍：2020/07/13周一
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 * todo Manacher算法
 */
class _647_PalindromicSubstrings {
    public int countSubstrings(String s) {
        if (s == null) return 0;
        int n = s.length();
        int ret = 0;
        for (int i = 0; i < 2 * n - 1; i++) {
            int l = i >> 1;
            int r = l + (i & 1);
            while (l >= 0 && r < n && s.charAt(l) == s.charAt(r)) {
                l--;
                r++;
                ret++;
            }
        }
        return ret;
    }

    class SolutionDP {
        public int countSubstrings(String s) {
            //f(i, j) 表示闭区间[i, j]的子串是否是回文串
            //f(i, j) = f(i + 1, j - 1) && s[i] == s[j]
            //i in [0, n - 1], j in [0, n - 1]
            int n = s.length();
            boolean[] dp = new boolean[n];
            int ret = 0;
            for (int i = n - 1; i >= 0; i--) {
                for (int j = n - 1; j >= i; j--) {
                    dp[j] = s.charAt(i) == s.charAt(j) && (j - i < 2 || dp[j - 1]);
                    if (dp[j]) {
                        ret++;
                    }
                }
            }
            return ret;
        }
    }
}

File: Week_06/kotlin/src/longestpalindromicsubstring/_647_PalindromicSubstrings.kt
package longestpalindromicsubstring

fun countSubstrings(s: String): Int {
    //f(i, j)表示s[i, j]是否是回文子串
    //f(i, j) = f(i + 1, j - 1) && s[i + 1] == s[j - 1]
    //i in [0, n - 1] j in [0, n - 1]
    var ret = 0
    val n = s.length
    val dp = BooleanArray(n)
    for (i in n - 1 downTo 0) {
        for (j in n - 1 downTo i) {
            dp[j] = s[i] == s[j] && (j - i < 2 || dp[j - 1])
            if (dp[j]) {
                ret++
            }
        }
    }
    return ret
}

class _647_Solution2_ {
    fun countSubstrings(s: String): Int {
        //0: 0 0,1: 0 1 ,2: 1 1,3: 1 2, 2n-2: n - 1 n - 1,
        val n = s.length
        var ret = 0
        for (i in 0..2 * (n - 1)) {
            var l = i shr 1
            var r = l + (i and 1)
            while (l >= 0 && r < n && s[l] == s[r]) {
                l--
                r++
                ret++
            }
        }
        return ret
    }
}